text/0194-mandatory-unique-identifier-for-sdk-based-telemetry-sources.md

```diff
@@ -0,0 +1,69 @@
+# Mandatory unique identifier for sdk-based telemetry sources
+
+Provide an explicit mandatory unique identifier for sdk-based telemetry sources.
+
+## Motivation
+
+Having a way to uniquely identify a telemetry source is helpful in many ways, like in processing and storing data from that source, visualizing them in a backend UI or debugging issues with that source and it's data.
+
+For sdk-based telemetry sources, as of now `service.name` (and related attributes `service.namespace` and `service.instance_id`) are the implicit standard for that due to `service.name` being enforced as mandatory by the [Resource SDK specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#sdk-provided-resource-attributes) and [Resource Semantic Conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md#semantic-attributes-with-sdk-provided-default-value).
+
+But, because those attributes are not **explicitly** available to uniquely identify a sdk-based telemetry source, multiple issues are calling out problems with the current state:
+
+* [opentelemetry-specification/issues#1034](https://github.com/open-telemetry/opentelemetry-specification/issues/1034) calls out that `service.instance.id` is poorly defined right now and should be replaced by something more meaningful, that can help to uniquely identify a SDK-based telemetry source.
+* [open-telemetry/opentelemetry-specification#2111](https://github.com/open-telemetry/opentelemetry-specification/pull/2111) calls out that there is no proper definition of what a `Service` is and that a proper definition is important since `service.name` is such an important attribute
+* [open-telemetry/opentelemetry-specification#2115](https://github.com/open-telemetry/opentelemetry-specification/pull/2115) asks for introducing `app.name` and others alongside `service.name` since client-side applications (browser, mobile) are **not** services and end-users might be confused by calling them a _Service_.
+* [open-telemetry/opentelemetry-specification#2192](https://github.com/open-telemetry/opentelemetry-specification/pull/2192)) is providing a middle ground between `app.name` and `service.name` by suggesting `telemetry.source.name` as broader term.
+
+To address all requirements outlined in those approaches, we are proposing the following combined approach for uniquely identifying a SDK-based telemetry source:
+
+* Introduce an `telemetry.sdk.source.id` attribute, which MUST either be autogenerated by the SDK at application start or be supplied via an environment variable to the SDK. This will be the unique identifier for an SDK-based telemetry.source.
```

Comment:
Why `telemetry.sdk.source.id` and not `telemetry.source.id`? How is "sdk" relevant to the description of the workload that is producing telemetry?

Reply:
It's one of my open questions: 
> Is the namespace `telemetry.sdk.source.*` suitable? Alternative names could be used: [...] `telemetry.source.*` as suggested by open-telemetry/opentelemetry-specification#2192. The difference is that[`telemetry.sdk.source`] does state explicitly that only SDK-based telemetry sources are covered. This is not necessarily bad, since other telemetry sources could decide to use it as well. [...]

Reply:
I would prefer to drop the word "source" an have `telemetry.sdk.id` (or `telemetry.sdk.instance.id`) to identify the SDK. This will match other `telemetry.sdk.*` attributes. This also fits the requirements of possible future remote SDK management capabilities, where the _SDK_ is actually the entity that is being managed.

To identify the Service we will (continue to) use `service.*` attributes, to identify other types of sources which are not Services we can introduce new sets of attributes (e.g. for client-side apps as was discussed earlier).

If we call this attribute `telemetry.source.id` it is not clear what does it identify. Does it identify the Service? If yes then do we remove `service.instance.id`? If we remove it then do we use a mix of `telemetry.source.id`+`service.*` attributes to identify the Service? This does not look very consistent to me.

Generally, for any particular _kind_ of entity that we want to identify I think we should have a set of attributes `<kind>.*` that are defined in semantic conventions specifically for that particular kind. This has been the practice so far and `telemetry.source.id` would be a conceptual deviation from the practice which I don't see why we need.

Reply:
@tigrannajaryan personally I am not completely sold on `telemetry.source.id` and I think it should be a different OTEP (i.e. assigning an opaque unique ID to a source vs. identifying the source via a collection of domain-specific attributes, like k8s.pod etc.) from the question of how to generalize `service.name` to different types of workloads.

I also find it more intuitive to have `service.name` and `app.name` vs. a single (and ambiguous) `telemetry.source.name`. The issue is how that affects existing backends like Jaeger where `service.name` is a core concept and `app.name` means nothing at all. Should the exporters to such backends translate `app.name` (OTEL) -> `service.name` (Jaeger)? This translation becomes easier / more consistent if we use a single `telemetry.source.name` instead of domain-specific attributes, but I agree that it diverges from most of the existing conventions.

Reply:
That's fine if this was only Jaeger, but from the comments in different threads it seems a lot of backends would prefer a single attribute. Maybe a fair question is: are there (how many) backends that would be completely fine with _not_ having a single attribute identifying the source by low cardinality name?

Reply:
That's a good question, we probably want vendors to chime in.

Just one data point: the AWS Xray exporter in Collector for example probes a series of different attributes (including but not limited to service.name) to arrive at a "name", see: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/af3d9b1dc9f1da2e4f71944a32e388368fb0f5e6/exporter/awsxrayexporter/internal/translator/segment.go#L131

I would expect each vendor to either have an equivalent logic that makes sense for their backend or perhaps we extract this as a common logic somewhere for all vendors to use.

Note that vendors likely need to be ready for this anyway since Collector doesn't necessarily set service.name for all data it collects/produces on its own.

Reply:
> Just one data point: the AWS Xray exporter in Collector for example probes a series of different attributes (including but not limited to service.name) to arrive at a "name"

I think this is a very good example, why there should be clarity on how to identify a telemetry source and why I opened my issues initially and now this OTEP: 

As of now it is really hard to say "this telemetry source is called <X>"? 

> I would expect each vendor to either have an equivalent logic that makes sense for their backend or perhaps we extract this as a common logic somewhere for all vendors to use.

That common logic is what I am looking for. My initial assumption was that `service.namespace`, `service.name` and `service.instance.id` are doing the job, but as we know now this has the following issues:
1. not everybody thinks of their workload as a "service"
2. it's not clear how `service.instance.id` is filled (see https://github.com/open-telemetry/opentelemetry-specification/issues/1034)

The intend of `telemetry.(sdk?).source.(id|name|namespace)` was to avoid that kind of lengthy checks we see in the AWS Xray exporter (@martinkuba brought up this argument here as well: https://github.com/open-telemetry/opentelemetry-specification/pull/2192)

An alternative, that was suggested at some point, is a `<whatever>.type` field  which is then `app` or `service` or ... and can help the implementation to jump to the correct attributes immediately.